DESIGN_PRINCIPLES.md

```diff
@@ -0,0 +1,24 @@
+# Concourse Design Principles
+
+Concourse's goal is to solve automation, once and for all, without becoming part of the problem.
+
+
+## Expressive by being precise
+
+Concourse should provide concepts that build a strong mental model for the user's project that remains intuitive as their automation requirements grow.
+
+Concepts should precisely outline their motivation and intended workflows. Friction and complexity resulting from the imprecise application of a concept should be a cue to introduce new ideas. ([Example](https://blog.concourse-ci.org/reinventing-resource-types/))
+
+
+## Versatile by being universal
+
+Concourse should be able to do a lot with a little. New concepts should only be introduced if their intended workflow cannot be precisely expressed in terms of existing concepts.
+
+Concepts should not be highly specialized for one domain or introduce tight coupling to specific technologies. Users should be able to relate to every concept, and their automation should be insulated from the constant churn of the tech industry.
+
+
+## Safe by being destructible
+
+Concourse should prevent [anti-patterns](https://github.com/concourse/concourse/wiki/Anti-Patterns) and the accumulation of technical debt. Concourse's concepts should make good practices feel intuitive and bad practices feel uncomfortable.
+
+Automation should be portable and reproducible. Recovery from disaster scenarios (e.g. total cluster loss) should be possible with minimal disruption to the path to production. Each piece of mission-critical state should be kept in an external source of truth, allowing Concourse to be an ephemeral choreographer.
```

Comment:
Is portability still a core tenet? I remember there were conversations about how even though Concourse tries to prevent vender lock in via separation of pipelines and scripts, it's still very difficult to rip Concourse out of an existing infrastructure.

In addition, I'm not sure how much this principle has actually affected RFCs, I can technically argue that any enhancements to pipelines violates portability since it'll make it harder for me switch off Concourse. It might better be expressed as a clean separation between infrastructure config (pipeline/task), and code (script).

Reply:
Portability in this context doesn't mean portability between Concourse and other CI systems - it just means the ability to easily migrate between Concourse clusters if needed, possibly running different versions (to the extent that needed features are supported) or being maintained by different operators with different configurations.

For example, this principle is what leads us to prefer that pipelines be as self-contained as possible, configuring their own dependent resource types rather than having them be configured cluster-wide.

Reply:
Ah, can we clarify that, maybe something like:

```suggestion
Automation should be portable and reproducible. Recovery from disaster scenarios (e.g. total cluster loss) should be possible with minimal disruption to the path to production. Each piece of mission-critical state should be kept in an external source of truth, allowing individual Concourse clusters to be an ephemeral choreographer.
```

Reply:
How's this?

```suggestion
Automation should be self-contained and reproducible in order to maintain business continuity when recovering from disaster scenarios (e.g. total cluster loss). Concourse should merely be a choreographer of mission-critical state kept in an external source of truth, allowing individual installations to be ephemeral.
```